inc/listeners/heartbeat.h

```diff
@@ -0,0 +1,91 @@
+#ifndef HEARTBEAT_H
+#define HEARTBEAT_H
+
+#include <iostream>
+#include <deque>
+#include <vector>
+#include <chrono>
+#include <fmt/chrono.h>
+
+#include "events.h"
+#include "instruction.h"
+
+class Heartbeat {
+public:
+  static constexpr auto cli_key = "Heartbeat";
+
+  uint64_t cycles_between_printouts = 10000000;
+  std::vector<uint64_t> num_retired_last_printout;
+  std::vector<uint64_t> cycles_last_printout;
+  std::vector<uint64_t> num_retired;
+  
+  std::vector<uint64_t> num_retired_start_phase;
+  std::vector<uint64_t> cycles_start_phase;
+  std::vector<bool> switched_phase; // true if there has been a begin_phase event since the last time RETIRE occurred
+
+  template<Event e, typename... Args>
+  void handle_event(Args&... args);
+  
+  void add_cpu(uint32_t cpu) {
+    while (cpu >= switched_phase.size()) {
+      num_retired_last_printout.push_back(0);
+      cycles_last_printout.push_back(0);
+      num_retired.push_back(0);
+      num_retired_start_phase.push_back(0);
+      cycles_start_phase.push_back(0);
+      switched_phase.push_back(false);
+    }
+  }
+};
+
+std::chrono::seconds elapsed_time();
+
+namespace heartbeat {
+
+template<Event e, typename... Args>
+inline void handle_event(Heartbeat* hb, Args&... args) {
+  //std::cout << "WARNING: generic handle event\n";
+}
+
+template<>
+inline void handle_event<Event::BEGIN_PHASE>(Heartbeat* hb, bool& is_warmup) {
+  for (size_t i = 0; i < hb->switched_phase.size(); i++) {
+    hb->switched_phase[i] = true;
+  }
+}
+
+template<>
+inline void handle_event<Event::RETIRE>(Heartbeat* hb, uint32_t& cpu, std::deque<ooo_model_instr>::const_iterator& begin, std::deque<ooo_model_instr>::const_iterator& end, uint64_t& current_cycles) {

```

Comment:
```suggestion
template<typename ITER>
inline void handle_event<Event::RETIRE>(Heartbeat* hb, uint32_t& cpu, ITER begin, ITER end, uint64_t& current_cycles) {

```
It might be good not to depend on the exact type of the iterator.

Reply:
I don't think we can do this, because partial specialization of function templates isn't allowed by C++.

Reply:
Blast! Foiled again... I suppose it isn't necessary, I was mostly trying to avoid depending on the type when we only use it for the distance between begin and end.